Feature: Add an extension popup to show cookie summary.

ayushnirwal opened the pull request:
## Description

Adds an pop up accessible by clicking on the popup icon. This pop up shows 2 piecharts ( one for 1p cookies and other for 3p ) giving information about what category of use they lie in.

<!-- What do we want to achieve with this PR? -->



## Screenshot/Screencast
![Screenshot 2023-07-03 at 7 42 49 PM](https://github.com/GoogleChromeLabs/ps-analysis-tool/assets/53055971/7ee4d535-77c3-4d73-9955-76930f8e93eb)



<!-- Please provide Screenshot/Screencast, if applicable -->

---

<!--
Example:

Fixes #123
Partially addresses #22
See #834
-->

Fixes #5

sayedtaqui commented:
@ayushnirwal 

1. Would you like to show the cookies count in this PR or later in a separate PR?
2. When we click on the already opened tab, can we at least show a message asking to refresh the page in v0.1 instead of not opening the popup?

<img width="622" alt="image" src="https://github.com/GoogleChromeLabs/ps-analysis-tool/assets/6297436/0ad7f0d7-36f6-4aab-b5d2-69f0ff2fb3f5">

sayedtaqui commented on packages/extension/src/utils/countCookiesByCategory.ts:
```diff
@@ -0,0 +1,86 @@
+
+  stats.total = cookieList.length;
+
+  for (const {
+    analytics,
+    parsedCookie: { domain },
+  } of cookieList) {
+    const _isFirstParty = isFirstParty(domain, tabUrl);
+
+    _isFirstParty ? stats.firstParty.total++ : stats.thirdParty.total++;
+
+    if (analytics) {
+      if (analytics.category.toLowerCase() === 'functional') {
+        _isFirstParty
+          ? stats.firstParty.functional++
+          : stats.thirdParty.functional++;
+      } else if (analytics.category.toLowerCase() === 'marketing') {
+        _isFirstParty
+          ? stats.firstParty.marketing++
+          : stats.thirdParty.marketing++;
+      } else if (analytics.category.toLowerCase() === 'analytics') {
+        _isFirstParty
+          ? stats.firstParty.analytics++
+          : stats.thirdParty.analytics++;
+      }
+    } else {
+      _isFirstParty ? stats.firstParty.unknown++ : stats.thirdParty.unknown++;
+    }
+  }
```
Can it be written something like this?

```js
  for (const cookie of cookieList) {
    const { analytics, parsedCookie: { domain } } = cookie;
    const isFirstPartyCookie = isFirstParty(domain, tabUrl);
    const category = analytics?.category?.toLowerCase();

    if (isFirstPartyCookie) {
      stats.firstParty.total++;
      stats.firstParty[category ? category : 'unknown']++;
    } else {
      stats.thirdParty.total++;
      stats.thirdParty[category ? category : 'unknown']++;
    }
  }
```

sayedtaqui commented on packages/extension/src/view/popup/index.html:
```diff
@@ -0,0 +1,14 @@
+  <head>
+    <title>CAT Popup</title>
+    <meta charset='UTF-8'>
+    <meta http-equiv='X-UA-Compatible' content='ie=edge'>
```
Any plans to run it in edge :D ?

sayedtaqui commented on webpack.config.cjs:
```diff
@@ -127,4 +127,26 @@ const devTools = {
+      color: '#fcd8ba',
+    }),
+    new HtmlWebpackPlugin({
+      title: 'CAT Popup',
```
```suggestion
      title: 'PSAT Popup',
```

sayedtaqui commented:
@ayushnirwal Just noticed that we are using the word "Uncategorised" in DevTool while we are using the word "Unknown" in the popup. Let's only use one of them.

sayedtaqui commented on packages/extension/src/view/popup/components/legend.tsx:
```diff
@@ -0,0 +1,52 @@
+    className="w-4 h-4 rounded-full flex items-center justify-center mr-2"
+    style={{ backgroundColor: color }}
+  >
+    <div className="w-2 h-2 rounded-full flex items-center justify-center bg-white"></div>
```
```suggestion
  <div className="w-2 h-2 rounded-full flex items-center justify-center bg-white" />
```

sayedtaqui commented on packages/extension/src/view/popup/components/legend.tsx:
```diff
@@ -0,0 +1,52 @@
+const Legend = ({ legendItemList }: LegendProps) => {
+  return (
+    <div className="flex flex-col">
+      {legendItemList.map(({ label, count, color }, idx) => (
```
```suggestion
      {legendItemList.map(({ label, count, color }, index) => (
```

sayedtaqui commented on packages/extension/src/view/popup/components/pieChart.tsx:
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import { VictoryPie } from 'victory';
+
+interface IPieChartProps {
```
```suggestion
interface PieChartProps {
```

sayedtaqui commented on packages/extension/src/view/popup/components/pieChart.tsx:
```diff
@@ -0,0 +1,46 @@
+          innerRadius={140}
+          animate={{ duration: 400 }}
+          data={data.map(({ count }) => ({ x: '', y: count }))}
+          labels={() => null}
```
```suggestion
          labels={() => ''}
```
My text editor shows that `label` is expecting a string or a number.

sayedtaqui commented on packages/extension/src/view/popup/index.html:
```diff
@@ -0,0 +1,12 @@
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <title>CAT Devtool</title>
```
```suggestion
    <title>PSAT DevTool</title>
```

ayushnirwal commented:
> @ayushnirwal
> 
> 1. Would you like to show the cookies count in this PR or later in a separate PR?
> 2. When we click on the already opened tab, can we at least show a message asking to refresh the page in v0.1 instead of not opening the popup?
> 
> <img alt="image" width="622" src="https://user-images.githubusercontent.com/6297436/250897411-0ad7f0d7-36f6-4aab-b5d2-69f0ff2fb3f5.png">

@sayedtaqui 
- I have added badge text in this PR since that is related to pop up.
- Also added messages "Please refresh the page"in the case you mentioned and "No Cookies" in the case if a tab does not use cookies.

sayedtaqui commented:
@ayushnirwal It shows "no cookies" instead of asking to refresh page on already loaded tab.

<img width="679" alt="image" src="https://github.com/GoogleChromeLabs/ps-analysis-tool/assets/6297436/a27c01ce-3e7e-4d57-adce-b411a5bf95d6">


---

and on a new tab

<img width="1440" alt="image" src="https://github.com/GoogleChromeLabs/ps-analysis-tool/assets/6297436/5acd2592-5220-4042-9a22-b68d3477711e">

sayedtaqui commented on packages/extension/src/view/popup/app.tsx:
```diff
@@ -0,0 +1,169 @@
+        </div>
+      ) : (
+        <div className="w-96 h-80 flex justify-center items-center flex-col">
+          <p className="font-bold text-lg">{'No cookies found'}</p>
```
```suggestion
          <p className="font-bold text-lg">No cookies found</p>
```

sayedtaqui commented on packages/extension/src/view/popup/const.ts:
```diff
@@ -0,0 +1,22 @@
+ */
+
+export const colourMap = {
+  funtional: '#00B04C',
```
```suggestion
  functional: '#00B04C',
```

sayedtaqui commented on packages/extension/src/view/popup/components/legend.tsx:
```diff
@@ -0,0 +1,52 @@
+  color: string;
+}
+
+const Circle = ({ color }: LegendIconProps) => (
```
We haven't kept more than one component in a single file. Let's create a separate file for it for consistency.

sayedtaqui commented on packages/extension/src/view/popup/components/legend.tsx:
```diff
@@ -0,0 +1,52 @@
+        <div key={index} className="w-36 flex items-center justify-center my-1">
+          <Circle color={color} />
+          <p className="flex-1 text-chart-label text-xs">{label}</p>
+          <p className=" text-chart-label text-xs"> {count} </p>
```
```suggestion
          <p className="text-chart-label text-xs">{count}</p>
```

sayedtaqui commented on packages/extension/src/view/popup/app.tsx:
```diff
@@ -0,0 +1,169 @@
+    tabURL: state?.url,
+  }));
+
+  if (!cookies) {
+    chrome.tabs.reload();
+  }
```
However the use `chrome.tabs.reload();` looks risky to me. What if a user has some work like they wrote some content in a textbox on their page and we reload the page? I would rather show a message asking to refresh themselves or click on a link to refresh the page.

sayedtaqui replied:
Let's create an issue for it to think about solving this problem later.

ayushnirwal replied:
I have remove the auto refresh for now.

sayedtaqui commented on packages/extension/src/localStore/tests/updateStorage.ts:
```diff
@@ -25,6 +25,11 @@ describe('local store: updateStorage', () => {
+      //@ts-ignore local does not implementations of other properties
+      action: {
+        // eslint-disable-next-line @typescript-eslint/no-empty-function
+        setBadgeText: async () => {},
```
Create `noop` util?

ayushnirwal replied:
It's not used anywhere else. We can create it when that happens

sayedtaqui commented on packages/extension/src/localStore/tests/updateStorage.ts:
```diff
@@ -25,6 +25,11 @@ describe('local store: updateStorage', () => {
 
   beforeAll(() => {
     globalThis.chrome = {
+      //@ts-ignore local does not implementations of other properties
```
typo

sayedtaqui commented on packages/extension/src/localStore/updateStorage.ts:
```diff
@@ -68,7 +68,18 @@ const updateStorage = async (
   }
-
+  try {
+    if (Object.keys(newStorageValue[tabId].cookies || {}).length > 0) {
+      await chrome.action.setBadgeText({
+        tabId: parseInt(tabId),
+        text: Object.keys(
+          newStorageValue[tabId].cookies || {}
+        ).length.toString(),
+      });
```
```suggestion
const tabCookies = newStorageValue[tabId].cookies || {};
const numCookies = Object.keys(tabCookies).length;

  if (numCookies > 0) {
    await chrome.action.setBadgeText({
      tabId: parseInt(tabId),
      text: numCookies.toString(),
    });
  }
```

sayedtaqui commented on packages/extension/src/utils/tests/countCookiesByCategory.ts:
```diff
@@ -0,0 +1,181 @@
+  headerType: 'request',
+};
+
+const emptyStats = {
```
Shall we use capital letters for all static constants like `EMPTY_STATS` in the project?

sayedtaqui commented on packages/extension/src/worker/service-worker.ts:
```diff
@@ -25,6 +25,9 @@ import {
 } from '../utils/fetchCookieDictionary';
 
+//clear data from previous sessions or version of the extention
+chrome.storage.local.clear();
```
Hmm, another clearing without any event listener when we have one on `chrome.runtime.onInstalled`? This may create confusion. Which event do we want to cover? Can we put it as part of their callbacks like `chrome.windows.onCreated `?

ayushnirwal replied:
I wanted to clear when the extension is refreshed ( not coverred by  )
